server/config.go

```diff
@@ -285,3 +294,161 @@ func (c *BalanceConfig) String() string {
+	defaultEtcdInitialClusterToken = "etcd-cluster"
+)
+
+func newEtcdConfig() *EtcdConfig {
+	return &EtcdConfig{
+		Name:                defaultEtcdName,
+		DataDir:             defaultEtcdDataDir,
+		WalDir:              defaultEtcdWalDir,
+		ListenPeerURL:       defaultEtcdListenPeerURL,
+		ListenClientURL:     defaultEtcdListenClientURL,
+		AdvertisePeerURL:    defaultEtcdAdvertisePeerURL,
+		AdvertiseClientURL:  defaultEtcdAdvertiseClientURL,
+		InitialCluster:      defaultEtcdInitialCluster,
+		InitialClusterToken: defaultEtcdInitialClusterToken,
+		InitialClusterState: embed.ClusterStateFlagNew,
+	}
+}
+
+func (c *EtcdConfig) clone() *EtcdConfig {
+	if c == nil {
+		return nil
+	}
+
+	ec := &EtcdConfig{}
+	*ec = *c
+	return ec
+}
+
+// String implements fmt.Stringer interface.
+func (c *EtcdConfig) String() string {
+	if c == nil {
+		return "<nil>"
+	}
+
+	return fmt.Sprintf("EtcdConfig(%+v)", *c)
+}
+
+// GenEmbedEtcd generates a configuration for embedded etcd.
+func (c *EtcdConfig) GenEmbedEtcd() (*embed.Config, error) {
+	cfg := embed.NewConfig()
+	cfg.Name = c.Name
+	cfg.Dir = c.DataDir
+	cfg.WalDir = c.WalDir
+	cfg.InitialCluster = c.InitialCluster
+	cfg.InitialClusterToken = c.InitialClusterToken
+	cfg.ClusterState = embed.ClusterStateFlagNew
+	cfg.EnablePprof = true
+
+	u, err := url.Parse(c.ListenPeerURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.LPUrls = []url.URL{*u}
+
+	u, err = url.Parse(c.ListenClientURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.LCUrls = []url.URL{*u}
+
+	u, err = url.Parse(c.AdvertisePeerURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.APUrls = []url.URL{*u}
+
+	u, err = url.Parse(c.AdvertiseClientURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.ACUrls = []url.URL{*u}
+
+	return cfg, nil
+}
+
+// Generate a unique port for etcd usage. This is only used for test.
+// Because initializing etcd must assign certain address.
+func freePort() int {
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		panic(err)
+	}
+
+	defer l.Close()
+
+	return l.Addr().(*net.TCPAddr).Port
+}
+
+// NewTestSingleEtcdConfig is only for test to create one etcd configuration.
+// Because pd-client also needs this, so export here.
+func NewTestSingleEtcdConfig() *EtcdConfig {
+	cfg := newEtcdConfig()
+	cfg.DataDir, _ = ioutil.TempDir("/tmp", "test_pd")
+	cfg.Name = "default"
+
+	clientPort := freePort()
+	peerPort := freePort()
+
+	cfg.ListenClientURL = fmt.Sprintf("http://localhost:%d", clientPort)
+	cfg.ListenPeerURL = fmt.Sprintf("http://localhost:%d", peerPort)
+
+	cfg.AdvertiseClientURL = cfg.ListenClientURL
+	cfg.AdvertisePeerURL = cfg.ListenPeerURL
+	cfg.InitialCluster = fmt.Sprintf("default=http://localhost:%d", peerPort)
+
+	return cfg
+}
+
+// NewTestMultiEtcdConfig is only for test to create multiple etcd configurations.
+// Because pd-client also needs this, so export here.
+func NewTestMultiEtcdConfig(count int) []*EtcdConfig {
```

Comment:
See above comment, pd-client should use this too, because initializing etcd config is too complex, and I don't want copy these codes in pd-client.